### DISCUSSION TITLE
Generator based zip

### DISCUSSION THREAD

**Discussion by bennidhamma** [remeda/remeda#970] [Suggest a Function]
I kind of like a generator based zip function that can take a few more parameters than just 2, is this something that would be a fit for remeda? I can't quite get my head around what it would look like as a data-last API.

```ts

export function* zip<T1, T2, T3, T4, T5>(
  A: T1[],
  B: T2[],
  C: T3[] = [],
  D: T4[] = [],
  E: T5[] = [],
): Generator<[T1, T2, T3, T4, T5]> {
  for (let i = 0; i < A.length; i++) {
    yield [A[i], B[i], C && C[i], D && D[i], E && E[i]];
  }
}

var colors = ["red", "blue", "green"];
var weights = [100, 200, 300];
var dates = [new Date(2021, 1, 1), new Date(2021, 1, 2), new Date(2021, 1, 3)];

// use with for...of
for (const [color, weight, date] of zip(colors, weights, dates)) {
  console.log(`${color.toUpperCase()}, ${weight / 100}, ${date.toTimeString()}`);
}

// or spread to map
console.log(
  [...zip(colors, weights, dates)].map(([c, w, d]) => ({c: c.charAt(0), w: w / 2, d})
);
```

I'm happy to polish it up and submit a PR if this is at all interesting.

**Comment by Bessonov** [comment]
From my point of view, var-args zip makes total sense, even it is unusual. But typings are challenging. And you can achieve similar result already now:

```typescript
var colors = ["red", "blue", "green"];
var weights = [100, 200, 300];
var dates = [new Date(2021, 1, 1), new Date(2021, 1, 2), new Date(2021, 1, 3)];

for (const [[color, weight], date] of zip(zip(colors, weights), dates)) {
	console.log(`${color.toUpperCase()}, ${weight / 100}, ${date.toTimeString()}`);
}

>>> RED, 1, 00:00:00 GMT+0100 (Central European Standard Time)
>>> BLUE, 2, 00:00:00 GMT+0100 (Central European Standard Time)
>>> GREEN, 3, 00:00:00 GMT+0100 (Central European Standard Time)
```

**Comment by cjquines** [comment]
possibly you could overload with different arg numbers, like how we type pipe

**Comment by eranhirsch** [comment]
I think adding a multi-zip function is a great idea.

It would be impossible to make a generic dataFirst and dataLast implementation, though, as we rely heavily on the runtime types (`typeof` operator) to decide what variant of the function is running now, and because everything is an array, there will be no way to distinguish between the versions.

We also can't change the signature of the existing function as it would break for a lot of users.

The 2 solutions I can think of are:
1. Adding `zip.<N>` flavors to zip (similar to how we do `indexed` and `strict`) and implement it for <N> in some range (3 to 6 maybe? 2 is already the default), so users can do `R.zip.4(arr1, arr2, arr3, arr4)` and also `R.pipe(..., R.zip.4(arr2, arr3, arr4) , ...)`
2. Adding a separate `multiZip` or `zipSeveral` (so they have the same prefix and are easily discoverable), or even `zip.n`, that would be dataFirst-only, and would take a variadic input, so users can do `R.zipSeveral(arg1, arg2, arg3, ..., argN)` or `R.pipe(..., ($) => R.zipSeveral($, arg2, arg3, ..., argN), ...)`

**Comment by eranhirsch** [comment]
@TkDodo - what are your thoughts here?

**Comment by cjquines** [comment]
hm, what if we make a function that takes a list of lists, and then zips that? the standard name would be `unzip`
